### PR TITLE
qos-scripts: fix uci callback handling

### PR DESCRIPTION
**Pull request by guidosarducci**
Maintainer: Felix Fietkau (@nbd168 )
Run tested: ar71xx/mips24kc under LEDE 17.01.4
(by comparing different versions' output of `generate.sh all` using the package default config)

Description:
Hi Felix,

This follows up a [request](https://github.com/openwrt/openwrt/pull/805#issuecomment-395434782) from @jow about a `qos-scripts` config parsing issue, and might also be relevant to this [report](https://forum.lede-project.org/t/anyone-having-issues-with-qos-app-on-f18f08d/15549) on the LEDE Forum. The root problem was in `qos-scripts` handling of start-of-section and end-of-section config parsing.

This should also be cherry-picked for `lede-17.01` and `openwrt-18.06` since those branches are identical.

Thanks,
Tony

### CODE REVIEW

**Comment by guidosarducci** [comment]
Pinging @hauke  also as this is something that should be backported for LEDE 17.01.5 if possible.

@jow- You didn't identify or link the problem report/reporter in your earlier [request ](https://github.com/openwrt/openwrt/pull/805#issuecomment-395434782). Could you ask them to confirm this fixes their issue?

Thank you both!

**Comment by dedeckeh** [comment]
@guidosarducci A separate PR https://github.com/openwrt/openwrt/pull/1084 has been opened for the same problem.
I asked @sch-m if this PR solves the observed issue but it seems not to be the case

**Comment by guidosarducci** [comment]
>dedeckeh commented 3 days ago
@guidosarducci A separate PR #1084 has been opened for the same problem.

Although deceptively simple, PR #1084 only partially uses the documented approach for config parsing and still relies on some broken behaviour/side-effects later fixed by 73d8a6ab, making it fragile. Using the default UCI config with 73d8a6ab does seem to work, but after reverting 73d8a6ab (i.e. stock 17.01.4 system) it hangs with an eventual OOM or SEGFAULT error.

> I asked @sch-m if this PR solves the observed issue but it seems not to be the case

OK, after confirming my fix still works in my test environment, I realized I dropped part of the changes from the git commit by mistake -- I've pushed an update now that works as advertised. Sorry 'bout that.

Do note that this change works independently of 73d8a6ab, so could be pushed out as a general `qos-scripts` update for 17.01 (even 17.01.4), 18.06 and master, which would allow 73d8a6ab to be reapplied later without breakage. Famous last words... :-)

**Comment by sch-m** [comment]
@guidosarducci @dedeckeh 
With the updated version the script now works as expected.

I am also testing with the default UCI config and an codebase without 73d8a6a.
I cannot reproduce ".. it hangs with an eventual OOM or SEGFAULT error."

The goal of my commit in PR #1084 was to fix the problem in a minimally invasive way.
In addition, I do not really like that the processing of one section is finished by the call of config_cb() for the next section. But that is only my personal opinion.

long story short: I am fine with both solutions, as long as the bug is fixed. :-)

@dedeckeh you can close my PR #1084 if you merge this one here.

**Comment by sch-m** [comment]
I've closed #1084.

I think we should use this one because @guidosarducci is right with his hint, that this solution "uses the documented approach for config parsing".

**Comment by guidosarducci** [comment]
I pushed an small update for the `qos-stat` script since it also uses callbacks as reported in [FS#1602](https://bugs.openwrt.org/index.php?do=details&task_id=1602).

@sch-m 
>I am also testing with the default UCI config and an codebase without 73d8a6a.
I cannot reproduce ".. it hangs with an eventual OOM or SEGFAULT error."

Did it work in the end? I wonder if it just failed in a different way? Those errors are what led me to notice that `$CONFIG_SECTION` was also being misused and could cause a problem.

>The goal of my commit in PR #1084 was to fix the problem in a minimally invasive way.
In addition, I do not really like that the processing of one section is finished by the call of config_cb() for the next section. But that is only my personal opinion.

I definitely agree with the minimally invasive approach in general. The way `qos-scripts` uses callbacks doesn't seem appropriate for the "stream processing" they were intended for, and overly complicates things. I dropped it from `qos-stat` and will try dropping it from `generate.sh`. It's more invasive, but may be simpler in the end. Large parts of the code are 12 years old, so maybe a little update is OK.

If you could test with your own custom configs I'd appreciate it, since I'm not a regular `qos-scripts` user.

**Comment by sch-m** [comment]
> I'm not a regular qos-scripts user.

Neither do I, so I have no custom config so far.

**Comment by guidosarducci** [comment]
A `qos-scripts` user in [FS#1602](https://bugs.openwrt.org/index.php?do=details&task_id=1602) has confirmed `qos-stat` now working properly. I've also asked them to verify the additional changes just pushed here, which use UCI callbacks only where appropriate, and non-callback parsing otherwise. The resulting code is smaller and simpler.

@sch-m Could you review as well?

**Comment by dedeckeh** [comment]
@guidosarducci Can you update the PKG_VERSION as well ?

**Comment by guidosarducci** [comment]
@dedeckeh I bumped the version to 1.3.1. Did you really mean PKG_RELEASE in fact?

Also, user `lantsem` in [FS#1602](https://bugs.openwrt.org/index.php?do=details&task_id=1602) has confirmed the last changes to simplify parsing work properly for them.

**Comment by sch-m** [comment]
@guidosarducci The newest version seems to handle the interface sections improperly. This iptables rules are missing compared to the former implementation (tested with the default qos config):
`-A FORWARD -o <WAN-IF> -j qos_Default`
`-A OUTPUT -o <WAN-IF> -j qos_Default`

**Comment by guidosarducci** [comment]
>**@sch-m commented an hour ago**
@guidosarducci The newest version seems to handle the interface sections improperly. This iptables rules are missing compared to the former implementation (tested with the default qos config):
`-A FORWARD -o <WAN-IF> -j qos_Default`
`-A OUTPUT -o <WAN-IF> -j qos_Default`

This seems familiar and IIRC looks like an error in your test setup, probably from using the **old** `generate.sh` and **new** `functions.sh` together. Make sure you're using both **old** `generate.sh` and **old**  `functions.sh` as your baseline test. Another red flag is the fact the default UCI config has `option enabled 0`, so one wouldn't expect those iptables rules to be present in the former setup.

For the record, I used 4 test cases after some discussion in [FS#1602](https://bugs.openwrt.org/index.php?do=details&task_id=1602) and noted the output of `generate.sh all | md5sum`. My "custom" UCI config used non-default `interface` device, `upload`, `download` and `enabled 1`.

**Baseline Test**
1. old `generate.sh` + old `/lib/functions.sh` + default `/etc/config/qos`  =>  71ae0373...
2. old `generate.sh` + old `/lib/functions.sh` + custom `/etc/config/qos`  =>  c26b557c...

**PR#1044 Test (73d8a6a backwards compatibility)**
1. new `generate.sh` + old `/lib/functions.sh` + default `/etc/config/qos`  =>  71ae0373...
2. new `generate.sh` + old `/lib/functions.sh` + custom `/etc/config/qos`  =>  c26b557c...

**PR#1044 Test (73d8a6a forward compatibility)**
1. new `generate.sh` + new `/lib/functions.sh` + default `/etc/config/qos`  =>  71ae0373...
2. new `generate.sh` + new `/lib/functions.sh` + custom `/etc/config/qos`  =>  c26b557c...

As you can see the rules all match as expected, and were also verified with another custom config in FS#1602.

**Comment by sch-m** [comment]
Sorry, my mistake. You are right. I forgot to enable the interface section in the default config in my last test.

**Comment by dedeckeh** [comment]
Patch pushed to my staging tree; thx

**Comment by dedeckeh** [comment]
@guidosarducci Patch will first be pushed into master; if no regression issues are reported I will backport both patches to 18.06. Regarding 17.01 I don't want to take any risk as only critical issues are backported to this branch
